MASSIF-1 Crystal Direct Harvester implementation

jbflo opened the pull request:
- This is the code related to the implementation of the Crystal Direct Harvester in Massif-1 ESRF

- While transforming the components related to the Harvester  to functional component , i also changed most of the Equipment Class to function.

- Apologize, I did not mean to make this PR that Big

Beamline that does not have an Harvester should not have any new things/modification to do
A PR in Mxcubecore related to this one need to be reviewed (first) as well .

![image](https://github.com/mxcube/mxcubeweb/assets/7275231/d9cb7c1f-386d-4e85-b324-b5d6d63fe1e6)

Commit:
Harvester Initial Config

Commit:
Add Harvester component, signal and routes

Commit:
Add Harvester reducers and Actions

Commit:
add view for Harvester  equipment and Maintenance

Commit:
Add image viewer and InOutSwitch button component

Commit:
Update Harvester and Maintenance Class equipment View

Commit:
dispatch harvester Contents to update view.

Commit:
Adjust Config

Commit:
Fix Mount from Harvester Logic

rename function

Commit:
Fix Mount From Harvester  logic  and component view

Fix Mount From Harvester signal init and component view

Commit:
FIx yaml indentation

Commit:
Add Image viewer component based on react-zoom-pan-pinch library

Commit:
ESlint rules and code  Flake8/ black formatting

Disabling C901 function is too complex (19)

Commit:
Small Fix

Commit:
Rebase, keep track of develop branch

Commit:
Add Doc-string and Type-hint format 

Function renamed

Commit:
improve Check whether mounting from Harvester or not

fabcor-maxiv commented on mxcubeweb/core/components/samplechanger.py:
```diff
@@ -467,66 +509,85 @@ def queue_mount_sample(view, data_model, centring_done_cb, async_result):  # noq
-                                "Could not center sample, skipping",
-                                "",
+            except Exception as ex:
+                print(str(ex))
```
I recommend using `some_logger.exception("Some message")`, it will automatically add the stack trace.

fabcor-maxiv commented on mxcubeweb/core/components/samplechanger.py:
```diff
@@ -467,66 +509,85 @@ def queue_mount_sample(view, data_model, centring_done_cb, async_result):  # noq
+                    except Exception:
+                        import traceback
+
+                        log.info("centring did not pass %s" % traceback.format_exc())
```
This is probably better:

```suggestion
                        log.exception("centring did not pass")
```

The exception including traceback should be automatically added: https://docs.python.org/3/library/logging.html#logging.exception

fabcor-maxiv commented on mxcubeweb/routes/harvester.py:
```diff
@@ -0,0 +1,111 @@
+            )
+            return resp
+
+        return jsonify(app.harvester.get_harvester_contents())
```
What is going on with `resp`? If we exclude the `except` block, `resp` is assigned value, twice, but never used. Did any linter catch that?

jbflo replied:
> What is going on with `resp`? If we exclude the `except` block, `resp` is assigned value, twice, but never used. Did any linter catch that?

This is Outdated .  was there when PR was WIP

Commit:
Use /api as endpoint

Commit:
Transform to functional component

Add css module

Commit:
Use state when render a Button as Label

Removed commented code

Commit:
Add function that were remove from previous commit

Commit:
Lint and small css adjust

Commit:
Small adjust on typo

Commit:
[Harvester] Python 3.9 compatibility

Remove Hard coded user Agent

fabcor-maxiv reviewed (commented):
Wow... I had pending review comments I had no idea about. Must have been from quite a while ago, so I have no idea if they are still meaningful.

fabcor-maxiv commented on ui/src/components/Equipment/SampleChanger.js:
```diff
@@ -1,313 +1,270 @@
-    );
+import styles from './equipment.module.css';
+
+// function SampleChangerTree(props) {
+//   let titleBackground;
+
+//   switch (props.state) {
+//     case 'READY': {
+//       titleBackground = 'success';
+
+//       break;
+//     }
+//     case 'MOVING': {
+//       titleBackground = 'warning';
+
+//       break;
+//     }
+//     case 'DISABLED': {
+//       titleBackground = 'default';
+
+//       break;
+//     }
+//     default: {
+//       titleBackground = 'danger';
+//     }
+//   }
+
+//   return (
+//     <Card className="mb-3" style={{ marginTop: '0.5em' }} bg={titleBackground}>
+//       <Card.Header>{props.title}</Card.Header>
+//       {props.children}
+//     </Card>
+//   );
+// }
```
Why?

jbflo replied:
This is unused and i was not sure why.
was not my code so Did not want to take the responsible to remove at first place .
I wanted to asked first what was the intention of it usage.  then i forgot

fabcor-maxiv commented on ui/src/components/Equipment/SampleChanger.js:
```diff
@@ -1,313 +1,270 @@
-    this.setState({ allow_control: !this.state.allow_control });
-    /* eslint-enable react/no-set-state */
-  }
+  // function itemClicked() {
+  //   setState({ allow_control: !state.allow_control });
+  // }
```
Why?

jbflo replied:
https://github.com/mxcube/mxcubeweb/pull/1145#discussion_r1551384820

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+                }
+                crystal_list.append(lst)
+        except Exception as ex:
+            print("Could not get Crystal List : %s" % str(ex))
```
Why no proper logging?

```suggestion
            some_logger.exception("Could not get Crystal List")
```

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+                crystal_list.append(lst)
+        except Exception as ex:
+            print("Could not get Crystal List : %s" % str(ex))
+            # TEMP return a fake list for test
```
Is it on purpose that we have test data here?

jbflo replied:
Yes, That's help test the GUI > TO be remove after

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+        return crystal_list
+
+    def get_global_state(self):
+        try:
+            return HWR.beamline.harvester_maintenance.get_global_state()
+        except Exception:
+            return "OFFLINE", "OFFLINE", "OFFLINE"
```
I really do not like those catch-all exceptions. There might be some actual issues somewhere in this block of code and we will never know about it, because it is swallowed and there is no logging! Is there really no cleaner way to handle this?

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+                    rest_token,
+                )
+            return True
+        except Exception as ex:
+            msg = "get all queue entries failed, reason:  %s" % str(ex)
+            logging.getLogger("user_level_log").exception(msg)
+            return False
```
There should be no need to have the exception in the message if it is the `exception` method of the logger that is used. This should be just fine:

```suggestion
        except Exception as ex:
            msg = "get all queue entries failed, reason"
            logging.getLogger("user_level_log").exception(msg)
            return False
```

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+
+        harvester_device.load_calibrated_pin()
+        harvester_device._wait_sample_transfer_ready(None)
+        print("waiting 40 seconds before mount")
+        # For some reason the Harvester return READY too soon
+        # approximately 40 Second sooner
+        gevent.sleep(40)
```
Why no proper logging?

Isn't there any condition that we can check to make sure we have waited long enough?

Why 40 seconds specifically? Is it empirical?

Should the 40 seconds value be made configurable? Or at the very least extracted into a named constant?

jbflo replied:
Yes empirical, this is something to deal with the device itself .
Until then , we shall stick with this. we 'll remove this line once they fix the issue on the device.

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+                ),
+            }
+
+            print(calibrated_motor_offset)
```
Is this here on purpose or should it have been deleted?

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+            )
+
+            harvester_device.set_calibrate_state(False)
+        except Exception as ex:
+            logging.getLogger("user_level_log").exception(
+                f"Pin Calibration / validation Failed {str(ex)}"
+            )
+            return False
```
Another catch-all-exceptions? Why? At least there is logging, though...


```suggestion
        except Exception:
            logging.getLogger("user_level_log").exception(
                "Pin Calibration / validation Failed"
            )
            return False
```

No, but seriously why not list the actual exceptions we want to catch here?

```
except (ExceptionFoo, ExceptionBar):
```

and of course, there should be much less coed in the `try` block. 

I know those things are not easy, but seeing code like this scares me. There is a very high risk of us staying naively unaware of a whole bunch of issues that are actually very easy to fix if we keep having this poor handling of exceptions.

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+            sample = self.get_sample_info(xtal_uuid)
+            self.app.sample_changer.mount_sample(sample)
+        except Exception:
+            return "Could not Harvest Crystal"
```
Maybe log the exception at least, and then maybe someone will be able to fix the issue if any shows up.

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+                sample_uuid, wait_before_load
+            )
+        else:
+            msg = harvester_device.get_status()
+            logging.getLogger("user_level_log").exception(
+                "ERROR: No Harvester Device Found"
+            )
+            logging.getLogger("user_level_log").exception(msg)
```
There is no exception here, we are not in an `except` block. So we should use the `error` method.

> `exception(msg, *args, **kwargs)`
>    Logs a message with level [ERROR](https://docs.python.org/3/library/logging.html#logging.ERROR) on this logger. The arguments are interpreted as for [debug()](https://docs.python.org/3/library/logging.html#logging.debug). Exception info is added to the logging message. This method should only be called from an exception handler.

-- https://docs.python.org/3/library/logging.html#logging.Logger.exception

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+        current_queue_dict = self.app.queue.queue_to_dict()
+        current_queue_list = list(current_queue_dict)
+        res = None
+        try:
+            res = current_queue_list.index(current_sample)
+        except (ValueError, IndexError):
+            res = None
```
That is better! : )

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+        """
+        current_queue = self.app.queue.queue_to_dict()
+        harvester_device = HWR.beamline.harvester
+        wait_before_load = (
+            False if harvester_device.get_room_temperature_mode() else True
+        )
```
Maybe can be simplified, no? I do not know for sure.

```suggestion
        wait_before_load = not harvester_device.get_room_temperature_mode()
```

fabcor-maxiv commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,505 @@
+        )
+        if harvester_device.get_number_of_available_pin() > 0:
+            gevent.sleep(2)
+            sample_UUID = current_queue[sample["sampleID"]]["code"]
```
```suggestion
            sample_uuid = current_queue[sample["sampleID"]]["code"]
```

No need to change casing here, is there?

fabcor-maxiv commented on mxcubeweb/core/components/samplechanger.py:
```diff
@@ -467,66 +509,85 @@ def queue_mount_sample(view, data_model, centring_done_cb, async_result):  # noq
+                            if not dm.current_centring_procedure:
+                                dm.start_centring_method(dm.C3D_MODE)
+
+                            # NBNB  BUG . self and app are not avialble here
```
???

fabcor-maxiv commented on ui/src/components/Equipment/Harvester.jsx:
```diff
@@ -0,0 +1,168 @@
+    </Menu>
+  );
+
+  const crystalUUID = props.contents.harvester_crystal_list;
```
```suggestion
  const crystalUuid = props.contents.harvester_crystal_list;
```

`crystalUuid` is fine, no need for extra upper case letters (at least in my opinion)

fabcor-maxiv commented on ui/src/components/Equipment/HarvesterMaintenance.jsx:
```diff
@@ -0,0 +1,179 @@
+    msg = message;
+  }
+
+  let plateBarcode = '';
+  if (global_state.plate_barecode) {
+    plateBarcode = global_state.plate_barecode.toString();
+  }
```
I think these things can be simplified to something like the following (and then we have `const`, which I think is preferrable):

```suggestion
  const plateBarcode = global_state.plate_barecode ? global_state.plate_barecode.toString() : "";
```

Also isn't there a formatting convention to always use either single quotes`'` or double quotes `"`?

jbflo replied:
Better like this: thanks
`const plateBarcode = global_state.plate_barecode?.toString();`

fabcor-maxiv reviewed (commented):
That is a lot of code!

I looked at some random spots (mostly the Python part).

I do not know enough about the inner workings of mxcube and/or harvester so my comments are mostly superficial. Mostly it is things that I believe a linter should catch, but I guess we must have very lenient linting rules (which is not so great for code quality in the long term). I hope one day I have time to look at those linting rules and maybe try to tighten things up in terms of code quality.

fabcor-maxiv commented:
I converted as draft. Maybe pull requests should stay as draft until they are done (or very close to be done) including the code passes all the Actions workflows (linting, formatting, tests cheking, and so on).

elmjag commented:
We need to make sure that all CI tests are passing before merging.

The CI tests are failing with this error:

    ModuleNotFoundError: No module named 'mxcubecore.HardwareObjects.Harvester'

I guess this PR requires https://github.com/mxcube/mxcubecore/pull/826 to be merged first. So, we need to sort out [826](https://github.com/mxcube/mxcubecore/pull/826) before merging this.

jbflo committed:
The exception including traceback should be automatically added

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

jbflo commented:
> We need to make sure that all CI tests are passing before merging.
> 
> The CI tests are failing with this error:
> 
> ```
> ModuleNotFoundError: No module named 'mxcubecore.HardwareObjects.Harvester'
> ```
> 
> I guess this PR requires [mxcube/mxcubecore#826](https://github.com/mxcube/mxcubecore/pull/826) to be merged first. So, we need to sort out [826](https://github.com/mxcube/mxcubecore/pull/826) before merging this.

Sure, As mentioned in the Description !

Commit:
Better handling of exception

passed url secret key

Commit:
Remove unused function and some Lint

jbflo commented:
> I converted as draft. Maybe pull requests should stay as draft until they are done (or very close to be done) including the code passes all the Actions workflows (linting, formatting, tests cheking, and so on).

@fabcor-maxiv This was in WIP  for quite some time waiting for reviews. 
But had to putted this has ready to catch peoples attention cause it's not so easy to review  your own PR when you know it's work. 

But Thanks a lot, don't hesitate to add your comment i'll address them as far as i can. 

This shall pass the  Checks, but we should get agree with @beteva First on [826](https://github.com/mxcube/mxcubecore/pull/826)  as @elmjag  Noticed

fabcor-maxiv commented:
> > I converted as draft. Maybe pull requests should stay as draft until they are done (or very close to be done) including the code passes all the Actions workflows (linting, formatting, tests cheking, and so on).
> 
> @fabcor-maxiv This was in WIP for quite some time waiting for reviews. But had to putted this has ready to catch peoples attention cause it's not so easy to review your own PR when you know it's work.

I see. I think you should be able to ask for code reviews from anyone in the top right corner of the page: 
![image](https://github.com/mxcube/mxcubeweb/assets/140523662/a60ad490-dde3-47ab-9e3e-366326b0d080)

Commit:
[Harvester] Remove temp  List ,

Get temp data from Mockup

Commit:
[Image Viewer] Add css module

Commit:
Move plate logic from Sample view  to  Plate Component

jbflo commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,481 @@
+                "centring_vertical": md.centringVertical.get_value(),
+            }
+
+            saved_position = md.saved_motor_position
```
@marcus-oscarsson Here for saved_motor_position

Commit:
check for AttributeError

Add use_harvester props in harvester component instead

Commit:
Remove unused component

jbflo commented on mxcubeweb/core/components/samplechanger.py:
```diff
@@ -183,6 +192,11 @@ def mount_sample_clean_up(self, sample):
         sc = HWR.beamline.sample_changer
 
+        try:
+            mount_from_harvester = sc.mount_from_harvester()
```
@marcus-oscarsson a review on this file will make the use of mount_from_harvester more clear

marcus-oscarsson replied:
Ok, thanks I think that some logic in `components/harvester.py` can be moved down to the harvester hardware object

marcus-oscarsson commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,494 @@
+        harvester_device.set_calibrate_state(False)
+        logging.getLogger("user_level_log").warning("Pin Calibration Canceled")
+
+    def validate_calibration(self) -> bool:
```
What do you think about moving this logic to the actual harvester hardware object, it seems mostly to interact with the diffractometer and the harvester and not so much with the UI

jbflo replied:
Maybe I should think of something else,  the harvester hardware object communicate with and only the Harvester . i would prefer not adding MD or other device interaction  to it.

but yes I/we should maybe not put too much on the mxcubeweb core components, and move things down to an Hardware object level.

marcus-oscarsson replied:
I think we will get alot of clarity from doing that.

jbflo replied:
Yeah Done.  Thanks !
And maybe that will help the guys from hamburg one day, if they ever want to have their Harvester in the Beamline :grin: . 

Have another look please, I hope this become more clear  :satisfied:

As we said maybe the mxcubeweb have old things to be move to mxcubecore level as well ? things like
-  specific queue logic , centering, mounting  etc..
 but for another story

marcus-oscarsson replied:
Yes that would be good, int that case we could make this an Adapter instead of a  Component, that would be much nicer.

marcus-oscarsson commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,494 @@
+        self.init_signals()
+        return self.get_harvester_contents()
+
+    def harvest_sample_before_mount(
```
This one could probably also with advantage be moved to the harvester object ?

jbflo replied:
This is just an interface , the method is implemented in  harvester object already .

Commit:
Remove get calibration function

Add api routes , deleted from previous commit if
Add possibility to send data collection to Crims (Button Click) if exist

Commit:
Move Function/logic from mxcubeweb to mxcubecore level

marcus-oscarsson commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,494 @@
+        return initial_state
+
+    def mount_from_harvester(self):
+        sc = HWR.beamline.sample_changer
```
You are not really gaining anything from this assignment other than more things on the stack :)

jbflo replied:
not really :sweat_smile: ,  I gained some line of code

That  help on having less try  except block 
when calling from other components. 
it only called from the sampleChanger component but twice , worth it.

marcus-oscarsson commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,494 @@
+            crystal_list = self.get_crystal_list()
+            room_temperature_mode = HWR.beamline.harvester.get_room_temperature_mode()
+            number_of_pins = HWR.beamline.harvester.get_number_of_available_pin()
+            contents = {
```
It would be nice to model contents, so that we know exactly what it returns. Like we have done in a few places, for instance here, https://github.com/mxcube/mxcubeweb/blob/develop/mxcubeweb/core/models/adaptermodels.py

jbflo replied:
I had a look, would be best for me making this into a new PR , if that's ok ?

marcus-oscarsson commented on mxcubeweb/core/components/harvester.py:
```diff
@@ -0,0 +1,494 @@
+    def get_calibration_state(self) -> bool:
+        return HWR.beamline.harvester.calibrate_state
+
+    def calibrate_pin(self) -> bool:
```
This can probably also be moved right ?

marcus-oscarsson commented on mxcubeweb/core/components/samplechanger.py:
```diff
@@ -135,6 +139,11 @@ def _addElement(parent, element):
 
             contents = {"name": root_name}
 
+            if hasattr(HWR.beamline.sample_changer, "get_room_temperature_mode"):
```
It would probably be better to add the `get_room_tempereature_mode` on the `AbstractSampleChanger`  that by default returns `False`

marcus-oscarsson commented on mxcubeweb/core/components/samplechanger.py:
```diff
@@ -183,6 +192,8 @@ def mount_sample_clean_up(self, sample):
 
         sc = HWR.beamline.sample_changer
 
+        mount_from_harvester = self.app.harvester.mount_from_harvester()
```
I guess we know when we are gong to mount from the harvester and we can then set `CENTERIN_ METHOD` to `manual` there ?

jbflo replied:
:worried: That can cause unexpected behavior 

Like this code for example : 
![image](https://github.com/mxcube/mxcubeweb/assets/7275231/f75996cd-ea9e-4ea0-a174-d6e50554156f)

we don't want the MANUAL CENTRING METHOD to be triggered also.

marcus-oscarsson replied:
I'm not sure exactly what code above refers to but `centring_method` should somehow refer to the setting set by the "global setting", so it should not be a problem (or we should make sure that its not :) )

marcus-oscarsson commented on mxcubeweb/core/components/samplechanger.py:
```diff
@@ -211,12 +222,17 @@ def mount_sample_clean_up(self, sample):
                     C3D_MODE = HWR.beamline.diffractometer.C3D_MODE
                     HWR.beamline.diffractometer.start_centring_method(C3D_MODE)
+                elif HWR.beamline.diffractometer.in_plate_mode():
+                    msg = "Starting autoloop Focusin ..."
```
typo -> "Starting autoloop Focusing ..."

marcus-oscarsson commented on mxcubeweb/core/components/samplechanger.py:
```diff
@@ -440,6 +457,11 @@ def queue_mount_sample(view, data_model, centring_done_cb, async_result):  # noq
 
+            # in this case the sample changer takes the sample from an Harvester
+            # We Harvest the sample and make it ready to load
+            if mount_from_harvester:
```
Can't we somehow know from the sample `data_model` that its a harvester sample, like this you dont need `mount_from_harvester`

marcus-oscarsson commented on mxcubeweb/core/components/samplechanger.py:
```diff
@@ -459,6 +481,10 @@ def queue_mount_sample(view, data_model, centring_done_cb, async_result):  # noq
                 )
 
+    # Harvest Next sample while loading current
+    if mount_from_harvester and not HWR.beamline.harvester.get_room_temperature_mode():
```
Its neater to do this with a signal on the Harvester `HardwareObject`, so that when a sample is mounted you trigger the harvesting of the following one.

Commit:
[Mxcubecore] Use correct version of mxcube core

Commit:
[pnpm] effect from Migrate to version 9

marcus-oscarsson commented:
Ok, so I'm merging this now and then we can follow up with the some improvements :)
